### ISSUE TITLE
3g$ will act as $ with `set nowrap`

### ISSUE THREAD

**Issue by icyhat** [vim/vim#4566]
1. without `set nowrap`, `3g$` acts OK.
2. with `set nowrap`, `3g$` acts as `$`.
3. `3$` works OK both cases while `3g$` works ERR.

**Comment by chrisbra** [comment]
Not sure what you mean. It seems to behave as [documented](https://vimhelp.org/motion.txt.html#g$):

```
g$ or g<End>            When lines wrap ('wrap' on): To the last character of
                        the screen line and [count - 1] screen lines downward
                        inclusive.  Differs from "$" when a line is wider
                        than the screen.
                        When lines don't wrap ('wrap' off): To the rightmost
                        character of the current line that is visible on the
                        screen.  Differs from "$" when the last character of
                        the line is not on the screen or when a count is used.
                        Additionally, vertical movements keep the column,
                        instead of going to the end of the line.
                        When 'virtualedit' is enabled moves to the end of the
                        screen line.
```

**Comment by icyhat** [comment]
> Not sure what you mean. It seems to behave as [documented](https://vimhelp.org/motion.txt.html#g$):
> 
> ```
> g$ or g<End>            When lines wrap ('wrap' on): To the last character of
>                         the screen line and [count - 1] screen lines downward
>                         inclusive.  Differs from "$" when a line is wider
>                         than the screen.
>                         When lines don't wrap ('wrap' off): To the rightmost
>                         character of the current line that is visible on the
>                         screen.  Differs from "$" when the last character of
>                         the line is not on the screen or when a count is used.
>                         Additionally, vertical movements keep the column,
>                         instead of going to the end of the line.
>                         When 'virtualedit' is enabled moves to the end of the
>                         screen line.
> ```

well well well, this document really need to be simplified...

1. `$` means go forward to a position of EOL(end of line), so `4$` will goto 4th EOL(end of line), right?, there is no need with `N-1` calculations....OMG...stupid docs...
2. what `g$` comes from is that `$` count a logic line as a line while `g$` count a wrap line(visual line) as a line, right?

here we go, if without `set nowrap`, `3g$` works as expected whether there is a wrap line or not, it go to the 3rd EOL

but with `set nowrap`, `3g$` only go forward to the 1st EOL, WHICH IS NOT AS EXPECTED..

**Comment by chrisbra** [comment]
> $ means go forward to a position of EOL(end of line), so 4$ will goto 4th EOL(end of line), right?, there is no N-1 calculations....OMG...stupid docs...

No, `N` is the count, so if you type `4$` it will go 3 lines down. That is because if you count the number of `EOL`, that is actually 4, 1 for the current line, plus for each of the following 3 lines 1. 

> what g$ comes from is that $ count a logic line as a line while g$ count a wrap line(visual line) as a line, right?

correct

> but with set nowrap, 3g$ only go forward to the 1st EOL,

That is in the documentation:

> When lines wrap ('wrap' on): To the last character of the screen line and [count - 1] screen lines downward inclusive.  Differs from "$" when a line is wider than the screen.

So when lines wrap, it will move the cursor down, but when lines do not wrap:

> When lines don't wrap ('wrap' off): To the rightmost character of the current line that is visible on the screen.

It doesn't mention here, that it will move the cursor a line down. It doesn't even mention what happens with a count. So this works as documented. 

> 3g$ only go forward to the 1st EOL, WHICH IS NOT AS EXPECTED..

Well, my guess is, it was never clear, what the count should actually mean, when the lines are not wrapped. Go down? By what? Screen lines or physical lines? Scroll horizontally to the next screen count times? So I guess it was ignored for the time being and nobody actually complained until today.

> WHICH IS NOT AS EXPECTED..

You haven't even mentioned one single time what would be expected. In the current state, it does what is documented (and therefore expected). Nothing more and nothing less

> well well well, this document really need to be simplified...

Well, how about proposing a document patch than, that makes it clearer? Or how about actually writing down what you would like `3g$` to move when lines are not wrapped? Looking at the initial bug report, it is quite terse and it is not clear, what needs to be fixed. The documentation? The behaviour? But how?

**Comment by icyhat** [comment]
1. > No, N is the count, so if you type 4$ it will go 3 lines down. That is because if you count the number of EOL, that is actually 4, 1 for the current line, plus for each of the following 3 lines 1.
    1. `$`: go **forward** to EOL(end of line)
    2. `[count] [motion]`: `N$`, go forward to EOL, do it 4 times, which is the result: go forward 4th EOL, anything wrong? when we learn `1 + 2 + 3`  first, we handle it with `1 + 2 + 3 = 3 + 3 = 6`, but why NO on `1 + 2 + 3 = 6`?
2. > That is in the documentation:
    1. vim updated itself since 1991, anything it updated indicating the past version's CONs, why documents can't be updated and  it's always absolutely right as a evidence to deny any changes?
    2. `g$` is to go forward on a visual EOF rather than a logic EOF, this is the only reason why it is invented, now `3g$` can work even in non-wrap lines which jump to 3rd visual EOF, but with `set nowrap`, it moved to 1st visual EOF, WHICH IS NOT EXPECTED
3. > Go down? By what? Screen lines or physical lines? 
    1. if not go down, why there is `N$` and `Ng$` supported already as go down without `set nowrap`?
    2. screen line? Hallelujah... `g$` is invented for visual line' EOL usage, of course it alwass move the cursor to the  rightest char in visual line, if not, why use `g$`, why not use `$`? 
4. > So I guess it was ignored for the time being and nobody actually complained until today.
    1. if not properly make it work, then why invent `Ng$`, just remove that `N`, make it like `^`, which can't use in `N^`
    2. nobody complained..OMG.. vim updated so many versions, are theses updates are all complained in 1991 and nothing new? every bug was pointed out first time, should not be the reason to deny who point it out.
5. > You haven't even mentioned one single time what would be expected.
    1. if `$` only mean current end of line, then remove `N` pattern on `$`, just like `0` or `^`
    2. if `N$` go next EOL N times, which why default is 1, documented in `quickref.txt`, then remove puzzled explanation `N-1 lines`, there is only `N`, and if reader wanna calculate `N-1` as they wish, and do support `$` will like `l`(right N chars) or `e`(right Nth end of word) to continue go on next position
    3. therefore `g$` will go Nth visual EOL forward, which as it is suppose to.
    4. what this issue comes out is that no consistency on `N$` with `Nl` or `Ne`, `Ne` can move forward to Nth word end acrossing lines, but `N$` cannot, and while other command explanation use `Nth`, this command use `N-1 lines`, this is terse.
6. >  Well, how about proposing a document patch than, that makes it clearer? Or how about actually writing down what you would like 3g$ to move when lines are not wrapped? Looking at the initial bug report, it is quite terse and it is not clear, what needs to be fixed. The documentation? The behaviour? But how?
    1. I'm willing to, but I can only do make docs better now since I know nothing of structure of vim code base.
    2. supposing that you are a experienced developer on vim, I guess it is easy to find how current `g$` jump to visual rightest char, copy that part code and recheck `set nowrap`, why different behaviour? 
    3. also other features, such as `$` as `e` can cross lines, this should be a added to `compatible options`, we can create a better vim , but since break compatibility, we give options, depending on user to enable or not the breaking compatible features

**Comment by icyhat** [comment]
in `quickref.txt`
```
|f|	N  f{char}	to the Nth occurrence of {char} to the right
|e|	N  e		forward to the end of the Nth word
```

but in `$`
```
|^|	   ^		to first non-blank character in the line
|$|	N  $		to the last character in the line (N-1 lines lower)
```
It uses `N-1 lines lower` which puzzled brains while other repeat N time, but this N - 1.

The truth is that like `e`, without parameter,  the default parameter is 1, so when we type `e`, it move forward to a position which is the end of a word. there is explanation as `N - 1 words righter`...It is `N`, move forward to a position whether it's a `end of word` or `end of line`, and added `N` means do it `N` times, which means forward N times, which Nth.

**Comment by chrisbra** [comment]
> [count] [motion]: N$, go forward to EOL, do it 4 times, which is the result: go forward 4th EOL, anything wrong? when we learn 1 + 2 + 3 first, we handle it with 1 + 2 + 3 = 3 + 3 = 6, but why NO on 1 + 2 + 3 = 6?

I have no idea, why you would think the counts should be added to 6. So no that is not clear why it should be 6 and this is not how other commands behave.

> vim updated itself since 1991, anything it updated indicating the past version's CONs, why documents can't be updated and it's always absolutely right as a evidence to deny any changes?

Documentation is changed all the time. However changing a command is not so easily possible, because users and scripts might depend on a certain behavior. 

> it moved to 1st visual EOF, WHICH IS NOT EXPECTED

You still haven't mentioned what would be expected. It behaves according to the documentation, so it certainly **is expected**


>    if not go down, why there is N$ and Ng$ supported already as go down without set nowrap?
>    screen line? Hallelujah... g$ is invented for visual line' EOL usage, of course it alwass move the cursor to the rightest char in visual line, if not, why use g$, why not use $?

Because `$` moves to the end of the line which is not what `g$` does. Note, that with nowrapping, since the visible lines do not wrap, it does not make sense to start moving the cursor by *physical* lines, since `g$` is supposed to move by *visible lines*. But since they do not wrap, I guess the count is silently ignored. 

> nobody complained..OMG.. vim updated so many versions, are theses updates are all complained in 1991 and nothing new? every bug was pointed out first time, should not be the reason to deny who point it out.

You have not yet proven, why the current behaviour is a bug. It behaves according to the documentation. If we change the behaviour now, it is a backward incompatible change. And the fact that nobody complained until now means, that apparently until now nobody considered the current behaviour a bug.

> if $ only mean current end of line, then remove N pattern on $, just like 0 or ^

Oh, some more incompatible changes, wonderful... 


> if N$ go next EOL N times, which why default is 1, documented in quickref.txt, then remove puzzled explanation N-1 lines, there is only N, and if reader wanna calculate N-1 as they wish, and do support $ will like l(right N chars) or e(right Nth end of word) to continue go on next position

I am not sure why you think that `N-1` is puzzling. It simply means, that if you use a count of 3 for the `$` command, the cursor will be moved 2 lines down, because that is where the *3rd* EOL occurs. That is quite clear. So it already behaves like your quoted `right N chars`, just written a bit differently.


> I'm willing to, but I can only do make docs better now since I know nothing of structure of vim code base.

To improve the documentation, you don't need to know the code, just current behaviour. 

> supposing that you are a experienced developer on vim, I guess it is easy to find how current g$ jump to visual rightest char, copy that part code and recheck set nowrap, why different behaviour?

Because as noted already, when nowrapping is set, there is no other visual line to move to.

**Comment by k-takata** [comment]
The document says `Differs from "$" when the last character of the line is not on the screen or when a count is used` when 'wrap' is not set, but it doesn't say how the count is used (or ignored).

The code is here:
https://github.com/vim/vim/blob/75a1a9415b9c207de5a29b25c0d1949c6c9c5c61/src/normal.c#L8042-L8050

Obviously it doesn't check the count (`cap->count1`).
I don't know whether this is a bug, but current implementation simply ignores the count.

**Comment by chrisbra** [comment]
Yeah, but I think this is, because in without wrapping, it is unknown, what the count actually means, since there are no visible lines to move to.

**Comment by k-takata** [comment]
I checked `test_normal.vim`. There is a test for `g$`, however, there are no tests with a count (and 'nowrap').

**Comment by icyhat** [comment]
## comment of your comment
there are 2 main ideas that your stand for
1. documentation defined xxx, yyy, zzz
2. incompatible

there are so many `cpoptions` already there, you can keep vim in compatible by default, but anything that is not reasonable should be updated to `cpoptions` as a choice offer to user to enable more consistent feature, just like it did in past years.
1. as treating `incompatible` features, neovim is more open and  `friendly`, they accept to face the ugly past and have courage to correct it because no one is perfect and nothing should pretend to be perfect. e.g. they add `cw` from `ce` to `cw` by `set cpooptions-=_` for consistency, and vim document gave its auto translate `cw` to `ce` as redundant needless explanation which is conflict with itself with `de` and `dw`. all these bugs should updated once we spent time on discussing for saving thousand other people's time in reading vim docs or using vim.

2. documentation is wrote by people, is not a unchangeable standard, not fully discussed, not final perfect definition, which is why we are discussing now, to make it better.
## consistency
1. `Ne`, go forward Nth EOW(End Of Word), default N is 1
2. `Nge`, go backward Nth EOW(End Of Word), default N is 1
Since this pattern is already there, I think `$` and `g$` should acts consistent pattern
1. `N$`, go forward Nth EOL(End Of Line), default N is 1
2. `Ng$`, go forward Nth visual EOL(End Of line), default N is 1
why need to think `N-1` while we we have a `N` simple explains what is `N`. if `N-1` is a better explanation, why not `e` use that explanation, `N-1` EOW(End Of Word) righter in documentations. It exist 2 ways to explain what `N` is, which is not simple and consistent, and not a better explanation using `N-1`, which is paying more brains on understand, and get nothing more.
```
quickref.txt

|$|	N  $		to the last character in the line (N-1 lines lower)
...
|e|	N  e		forward to the end of the Nth word
```
## what I thought bug is 
what vim broke the consistency now is that 
1. `Ng$` works as forward Nth EOL
2. `Ng$` works as forward 1st EOL with `set nowrap`
It doesn't make any reason to do like this even some old docs explains and explains very redundant and not consistent which is wasting others time.

now you tell me the documentation explains unnecessary branched of situation to defined what `g$` behaves, and it's as expected.

We can't think it should be like this just because it's written in book or some docs.

It's true if it's true by facts, whether in doc or anywhere.
## comparison
```
g$ or g<End>            When lines wrap ('wrap' on): To the last character of
                        the screen line and [count - 1] screen lines downward
                        inclusive.  Differs from "$" when a line is wider
                        than the screen.
                        When lines don't wrap ('wrap' off): To the rightmost
                        character of the current line that is visible on the
                        screen.  Differs from "$" when the last character of
                        the line is not on the screen or when a count is used.
                        Additionally, vertical movements keep the column,
                        instead of going to the end of the line.
                        When 'virtualedit' is enabled moves to the end of the
                        screen line.
```
1. docs using 12 lines text to explain an original single meaning is what stupid and waste thousand of people's time
2. they should be simplified as 
```
g$ or g<End>            Go forward to a visible EOL(End Of Line) on current screen.
                                   e.g.
                                       "4g$" will go forward to 4th visible EOL.
                                           a visible EOL count as a end of line show on screen
                                   comparison:
                                       "4$" will go forward to 4th logic EOL
                                           a logic EOL count as a char before the symbol of carriage return in text file
``` 
## logic "C-R" in text and visible "C-R" on screens
![omg](https://user-images.githubusercontent.com/42270578/59852409-52d1a700-93a1-11e9-96f4-144a9e30ad77.png)

click to see bigger picture

**Comment by chrisbra** [comment]
I am just commenting on one specific part of your last answer and leave the rest for Bram to decide what to do.

> there are so many cpoptions already there, you can keep vim in compatible by default, but anything that is not reasonable should be updated to cpoptions as a choice offer to user to enable more consistent feature, just like it did in past years.

Yes and we are already running out of letters for it. So adding another letter needs to be well-thought.

> as treating incompatible features, neovim is more open and friendly, they accept to face the ugly past and have courage to correct it because no one is perfect and nothing should pretend to be perfect. e.g. they add cw from ce to cw by set cpooptions-=_ for consistency, and vim document gave its auto translate cw to ce as redundant needless explanation which is conflict with itself with de and dw. all these bugs should updated once we spent time for saving thousand other people's time in reading vim docs or using vim.

That is a bit unfair to say. First, I believe it was actually my patch that was merged by Neovim folks. Second regarding this actual issue here, it was actually delegated by the Neovim people (because it was thought to be a bug) and let us handle it. Nobody discussed how it should actually behave for non-wrapped lines. 

Also, if you want to contribute and enhance documentation updates, simply post a patch and write down why you think the existing documentation needs improvements ;)

I might actually have a look at this and implement it, but before starting to work on, I want to make sure, that the proposed new behaviour is reasonable.

And please, if you want to make actual enhancements/changes to existing beahaviour, next time please fill out the issue template. After all there is a reason it exists and do not just write a very terse bug report (that is very hard to understand). 

Thanks.

**Comment by jamessan** [comment]
> Yeah, but I think this is, because in without wrapping, it is unknown, what the count actually means, since there are no visible lines to move to.

It seems reasonable to me that "visible line" = "physical line" in this scenario.

**Comment by icyhat** [comment]
there are only 2 aspects 
1. for computer store information => `$`, go forward Nth end of line, which is same with LINE NUMBER calculate, which count a `\n` as a EOL, but since we operate on char rather than cursor, highlighted to last position of char, just like empty line, highlighted first position of a char, even no chars in this line, if MOTIONS without COUNT, default COUNT is 1
2. for human eye and brains => `g$`, go forward Nth visible end of line, which count by human's eye moving on screen, if parameter  COUNT == null from user, default COUNT is 1 by function

these two aspects are clear and reasonable and consistent and easy to understand. Also I wanna offer help on docs, but since current Vim doesn't run as above logic, it runs now in below
1. `e` => forward to end of word, press `e` many times will go forward crossing line to end of word
2. `$` => only current line, press `$` many times will stay on end of line, not crossing line

this is why I think `$` should be consistent with 1 or 2
1. to be same with `0` or `^` which can't use N, and then its behavior in above easy understood
2. to be same with `e` or `w` which go forward to a position N times

also there is tricky when `j` after `$`, it will go end of line, which is not origin meaning, `j` should be a simple downward the cursor as possible(as possible means if downward has a char, go there, if have no char, go closest to that position which is the end of next line), which `$` is the key to go forward end of line. So when we move cursor to the end of current line, if we wanna downward, we use `j`, if we wanna next end of line, we use `$`.

also there is tricky why `_` can use N pattern to as `$` but `^` can't use N pattern

since I found this is tricky and not consistent, i canceled `set nowrap` in config file. it's now perfect only using `$` except multi `$` cannot cross line.

at initial, I just made a `3g$` and found that not consistent with my brains, without your reply, I never think of SCREEN and file and EOL and EOW  and default parameter is 1 and cursor and char.

Thanks : )
